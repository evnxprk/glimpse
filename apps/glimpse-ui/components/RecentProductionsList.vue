<template>
  <div class="past-prod-list">
    <VSkeletonLoader
      v-for="n in 10"
      :index="n"
      :key="n"
      type="card, list-item-two-line"
      class="past-prod-card"
      boilerplate
    />
  </div>
</template>

<script>
import gql from 'graphql-tag'

export default {
  name: 'RecentProductionsList',
  apollo: {
    productions: gql`query { productions {
        name
        id
}}`
  }
}
</script>

<style lang="scss" scoped>
  .past-prod-list {
    white-space: nowrap;
    overflow-x: scroll;
    .past-prod-card {
      display: inline-block;
      width: 300px;
      margin-right: 20px;
    }
  }
</style>
