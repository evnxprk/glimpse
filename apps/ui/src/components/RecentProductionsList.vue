<template>
  <div style="display: flex; justify-content: center;" class="mt-5">
    <div class="card-container">
      <div style="display: flex; justify-content: center" v-for="production in productions">
        <ProductionCard :id="parseInt(production.id)" class="mt-5"
            :start-time="new Date(production.startTime)" :description="production.description" :name="production.name" :thumbnailUrl="production.thumbnail?.path" />
      </div>
    </div>
  </div>
  <div class="link">
    <router-link to="productions" style="text-decoration: none">
      <v-btn variant="outlined" class="text-caption">See More...</v-btn>
    </router-link>
  </div>
</template>
<script setup lang="ts">
import ProductionCard from "@/components/ProductionCard.vue";
import type {Production} from "@/graphql/types";
import type {PropType} from "vue";

defineProps({
  productions: {
    type: Object as PropType<Production[]>,
    required: true
  }
})
</script>

<style scoped lang="scss">
.card-container {
  display: flex;
  justify-content: space-around;
  flex-wrap: wrap;
  max-width: 800px;
  @media (max-width: 800px) {
    max-width: 500px;
    justify-content: center;
  }
}
.productions {
  display: flex;
  justify-content: center;
}
.link {
  display: flex;
  justify-content: center;
  margin: 10px;
}

</style>
