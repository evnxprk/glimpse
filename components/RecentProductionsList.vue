<template>
  <div class="recent-prod">
    <VRow>
      <VCol v-for="n in 2" :key="n" sm="6">
        <span v-if="productions[n]">
          <NuxtLink :to="'/productions/' + productions[n].id" class="recent-prod-link">
            <VImg
              :src="productions[n].thumbnail.link"
              :title="productions[n].name"
              :aspect-ratio="16/9"
              gradient="to top, rgba(0,0,0,0.3), rgba(0,0,0,0)"
            />
          </NuxtLink>
        </span>
      </VCol>
    </VRow>
    <VRow>
      <VCol v-for="n in 3" :key="n" sm="3">
        <span v-if="productions[n+2]">
          <NuxtLink :to="'/productions/' + productions[n+2].id" class="recent-prod-link">
            <VImg
              :src="productions[n+2].thumbnail.link"
              :title="productions[n+2].name"
              :aspect-ratio="16/9"
              gradient="to top, rgba(0,0,0,0.3), rgba(0,0,0,0)"
            />
          </NuxtLink>
        </span>
      </VCol>
      <VCol sm="3">
        <NuxtLink to="/productions" class="recent-prod-link">
          <VResponsive :aspect-ratio="16/9" class="view-more">
            <h3>View More</h3>
          </VResponsive>
        </NuxtLink>
      </VCol>
    </VRow>
  </div>
</template>

<script>
import gql from 'graphql-tag'
import moment from 'moment'
import 'simplebar/dist/simplebar.min.css'

export default {
  name: 'RecentProductionsList',
  components: {
  },
  data () {
    return {
      productions: []
    }
  },
  methods: {
    getFormattedDate (production) {
      return moment(production.startTime).format('MMM Do YYYY')
    }
  },
  apollo: {
    productions: {
      prefetch: false,
      variables: {
        count: 6
      },
      query: gql`query RecentProductionsListProductions($count: Int!) {
        productions(pageSize: $count) {
          id
          description
          name
          thumbnail {
              name
              link
          }
          startTime
        }
      }`
    }
  }
}
</script>

<style lang="scss" scoped>
  .view-more {
    text-align: center;
    display: flex;
    align-items: center;
    background-color: #00000030;
    font-size: 1vw;
    @media (max-width 2500px) {
      font-size: 1.5vw;
    }
    @media (max-width: 600px) {
      font-size: 3vw;
    }
  }
  .recent-prod-link {
    color: white;
    text-decoration: none;
  }
</style>
