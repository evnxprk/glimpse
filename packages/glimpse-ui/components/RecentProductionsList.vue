<template>
  <simplebar ref="pastProdList" class="past-prod-list">
    <VCard
      :href="'/productions/' + prod.id"
      v-for="prod in productions"
      :index="prod.id"
      :key="prod.id"
      :nuxt="true"
      class="past-prod-card"
    >
      <VImg :src="prod.thumbnail.link" :title="prod.thumbnail.name" height="200px" />
      <div class="prod-card-text-wrapper">
        <VCardTitle>
          {{ prod.name }}
        </VCardTitle>
        <VCardSubtitle>
          {{ prod.description }}
        </VCardSubtitle>
      </div>
      <VRow>
        <VCol>
          <VCardText class="v-card__subtitle">
            {{ getFormattedDate(prod) }}
          </VCardText>
        </VCol>
        <VCol>
          <VCardActions class="prod-actions">
            <VBtn :href="'/productions/' + prod.id" text color="primary">
              Watch
            </VBtn>
          </VCardActions>
        </VCol>
      </VRow>
    </VCard>

    <VBtn
      v-if="showScrollButton"
      @click="pushLeft"
      fab
      dark
      x-large
      color="primary"
      class="scroll-btn"
      title="Click to scroll left"
      aria-label="Click to scroll left"
    >
      <font-awesome-icon :icon="['fal','chevron-right']" size="lg" />
    </VBtn>
  </simplebar>
</template>

<script>
import gql from 'graphql-tag'
import moment from 'moment'
import simplebar from 'simplebar-vue'
import 'simplebar/dist/simplebar.min.css'

export default {
  name: 'RecentProductionsList',
  components: {
    simplebar
  },
  data () {
    return {
      productions: [],
      showScrollButton: true
    }
  },
  mounted () {
    this.$refs.pastProdList.SimpleBar.getScrollElement().addEventListener('scroll', this.handleScroll)
  },
  beforeDestroy () {
    this.$refs.pastProdList.SimpleBar.getScrollElement().removeEventListener('scroll', this.handleScroll)
  },
  methods: {
    getFormattedDate (production) {
      return moment(production.startTime).format('MMM Do YYYY')
    },
    handleScroll (evt) {
      this.showScrollButton = (evt.target.scrollLeft === 0)
    },
    /**
     * Push this list left (or right, depending on how you think about it) one full width.
     * This exists for the right-chevron button, to demonstrate that this list is scrollable.
     * It scrolls in an ease-out manner one full width, or until the end of the list is hit.
     */
    pushLeft () {
      let pxToPush = this.$refs.pastProdList.SimpleBar.getScrollElement().offsetWidth // Scroll the width of the element
      // Runs every 5ms
      const interval = setInterval(() => {
        const pushAmount = pxToPush / 35 // Scroll by 1/35th of the remaining amount - Creates ease-out effect
        this.$refs.pastProdList.SimpleBar.getScrollElement().scrollLeft += pushAmount

        pxToPush -= pushAmount

        if (pxToPush <= 1) { // Stop when 1 px left, with 0 it breaks as you infinitely approach 0
          clearInterval(interval)
        }
      }, 5)
    }
  },
  apollo: {
    productions: gql`query RecentProductionsListProductions { productions {
        id
        description
        name
        thumbnail {
            name
            link
        }
        startTime
}}`
  }
}
</script>

<style lang="scss" scoped>
  @use "sass:map";
  @import "~vuetify/src/styles/styles.sass";

  .past-prod-list {
    position: relative;
    @media(min-device-width: 500px) {
      white-space: nowrap;
      overflow-x: scroll;

      -ms-overflow-style: none;
      &::-webkit-scrollbar {
        display: none;
      }

      .past-prod-card {
        margin-right: 20px;
      }
    }

    @media (max-device-width: 500px) {
      text-align: center;
      .past-prod-card {
        text-align: initial;
        margin-bottom: 20px;
      }
      .scroll-btn {
        display: none;
      }
    }
    .past-prod-card {
      display: inline-block;
      width: 300px;
      margin-bottom: 20px;
    }

    .prod-actions {
      float: right;
    }
  }
  .prod-card-text-wrapper {
    height: 150px;
    overflow: hidden;
    text-overflow: ellipsis;

    &:before {
      content:'';
      width:100%;
      height: 12%;
      position:absolute;
      left:0;
      bottom: 78px;
      background:linear-gradient(transparent, map.get($grey, "darken-3"));
    }
  }
  .scroll-btn {
    opacity: 0.7;
    &:hover {
      opacity: 1;
    }
    position: absolute;
    z-index: 100;
    right: 20px;
    top: 50%;
    transform: translateY(-50%);
  }
</style>
