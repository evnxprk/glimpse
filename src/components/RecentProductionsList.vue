<template>
  <n-grid cols="350:1 650:2 950:3 1250:4 1550:5 1850:6" y-gap="10" :collapsed="true" :collapsed-rows="2">
    <n-grid-item v-for="production in response.result.value?.findManyProduction" class="productions">
      <ProductionCard
        :id="parseInt(production.id)" :start-time="new Date(production.startTime)" :description="production.description" :name="production.name" />
    </n-grid-item>
  </n-grid>
  <div class="link">
    <router-link to="productions" style="text-decoration: none"><NButton>See more...</NButton></router-link>
  </div>
</template>

<script setup lang="ts">
import ProductionCard from "@/components/ProductionCard.vue";
import { NButton, NGrid, NGridItem } from "naive-ui";
import { useQuery } from "@vue/apollo-composable";
import { FindRecentProductionsDocument } from "@/graphql/types";

const response = useQuery(FindRecentProductionsDocument, {
  paginationInput: {
    take: 12
  }
})

</script>

<style scoped lang="scss">
.productions {
  display: flex;
  justify-content: center;
}
.link {
  display: flex;
  justify-content: center;
  margin: 20px;
}

</style>
