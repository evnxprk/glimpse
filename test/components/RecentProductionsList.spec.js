import { mount, RouterLinkStub } from '@vue/test-utils'
import Vue from 'vue'
import Vuetify from 'vuetify'
import moment from 'moment'
import VueRouter from 'vue-router'
import VuetifyOptions from '../../vuetify.config'
import RecentProductionsList from '~/components/RecentProductionsList'

Vue.use(Vuetify)
const vuetifyInstance = new Vuetify(VuetifyOptions)

describe('RecentProductionsList', () => {
  let wrapper
  const data = {
    productions: [
      {
        id: 0,
        description: 'Sample Description #0',
        name: 'Production #0',
        thumbnail: {
          name: 'Image for Production #0',
          link: 'https://pbs.twimg.com/media/EMnvt2bXUAAYQSL.jpg'
        },
        startTime: new Date('2013-02-03 01:20')
      },
      {
        id: 23,
        description: 'Sample Description #23',
        name: 'Production #23',
        thumbnail: {
          name: 'Image for Production #23',
          link: 'https://media.gettyimages.com/photos/man-impersonating-a-crossdressing-adolf-hitler-peeling-potoatoes-picture-id527186738'
        },
        startTime: new Date('2015-02-03 01:20')
      },
      {
        id: 10,
        description: 'Sample Description #10',
        name: 'Production #10',
        thumbnail: {
          name: '',
          link: 'https://media.gettyimages.com/photos/cracked-up-picture-id499219315?s=2048x2048'
        },
        startTime: new Date('2014-02-03 01:20')
      }
    ]
  }

  function mountFn (opts) {
    wrapper = mount(RecentProductionsList, {
      ...opts,
      vuetify: vuetifyInstance,
      router: new VueRouter(),
      stubs: {
        NuxtLink: RouterLinkStub
      }
    })
    wrapper.setData(data)
  }

  test('Properly renders', async () => {
    mountFn()
    await wrapper.vm.$nextTick()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('Contains correct data', async () => {
    mountFn()
    await wrapper.vm.$nextTick()
    const cards = wrapper.findAll('.past-prod-card')

    for (let i = 0; i < cards.length; i++) {
      const card = cards.at(i)
      expect(card.find('.past-prod-card').attributes('href'))
        .toEqual('/productions/' + data.productions[i].id)
      expect(card.find('.v-card__title').text()).toEqual(data.productions[i].name)
      expect(card.find('.v-card__subtitle').text()).toEqual(data.productions[i].description)
      expect(card.find('.v-card__text.v-card__subtitle').text())
        .toEqual(moment(data.productions[i].startTime).format('MMM Do YYYY'))
    }
  })

  test('Scroll hides move button', async () => {
    mountFn()
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.scroll-btn').exists()).toBe(true)
    wrapper.find('.simplebar-content-wrapper').element.scrollLeft = 100
    wrapper.find('.simplebar-content-wrapper').trigger('scroll')
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.scroll-btn').exists()).toBe(false)
  })

  // JSDOM doesn't like scrolling. This test doesn't appear to be working, should probably
  // be an end-to-end test.
  // test('Clicking scroll btn scrolls', async () => {
  //   mountFn()
  //   await wrapper.vm.$nextTick()
  //   expect(wrapper.find('.simplebar-content-wrapper').element.scrollLeft).toEqual(0)
  //   await wrapper.find('.scroll-btn').trigger('click')
  //   await new Promise((resolve) => {
  //     setTimeout(() => {
  //       resolve()
  //     }, 1000)
  //   })
  //   await wrapper.find('.simplebar-content-wrapper').trigger('scroll')
  //   expect(wrapper.find('.simplebar-content-wrapper').element.scrollLeft).toBeGreaterThan(0)
  // })
})
