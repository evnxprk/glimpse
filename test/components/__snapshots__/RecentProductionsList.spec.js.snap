// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentProductionsList Properly renders 1`] = `
<div
  class="recent-prod"
>
  <div
    class="row"
  >
    <div
      class="col-sm-6 col"
    >
      <span>
        <a
          class="recent-prod-link"
        >
          <div
            class="v-responsive v-image"
            title="Production #0"
          >
            <div
              class="v-responsive__sizer"
              style="padding-bottom: 56.25%;"
            />
            <div
              class="v-image__image v-image__image--preload v-image__image--cover"
              style="background-position: center center;"
            />
            <div
              class="v-responsive__content"
            />
          </div>
        </a>
      </span>
    </div>
    <div
      class="col-sm-6 col"
    >
      <span>
        <a
          class="recent-prod-link"
        >
          <div
            class="v-responsive v-image"
            title="Production #23"
          >
            <div
              class="v-responsive__sizer"
              style="padding-bottom: 56.25%;"
            />
            <div
              class="v-image__image v-image__image--preload v-image__image--cover"
              style="background-position: center center;"
            />
            <div
              class="v-responsive__content"
            />
          </div>
        </a>
      </span>
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col-sm-3 col"
    >
      <span>
        <a
          class="recent-prod-link"
        >
          <div
            class="v-responsive v-image"
            title="Production #10"
          >
            <div
              class="v-responsive__sizer"
              style="padding-bottom: 56.25%;"
            />
            <div
              class="v-image__image v-image__image--preload v-image__image--cover"
              style="background-position: center center;"
            />
            <div
              class="v-responsive__content"
            />
          </div>
        </a>
      </span>
    </div>
    <div
      class="col-sm-3 col"
    >
      <span>
        <a
          class="recent-prod-link"
        >
          <div
            class="v-responsive v-image"
            title="Production #189"
          >
            <div
              class="v-responsive__sizer"
              style="padding-bottom: 56.25%;"
            />
            <div
              class="v-image__image v-image__image--preload v-image__image--cover"
              style="background-position: center center;"
            />
            <div
              class="v-responsive__content"
            />
          </div>
        </a>
      </span>
    </div>
    <div
      class="col-sm-3 col"
    >
      <span>
        <a
          class="recent-prod-link"
        >
          <div
            class="v-responsive v-image"
            title="Production #76"
          >
            <div
              class="v-responsive__sizer"
              style="padding-bottom: 56.25%;"
            />
            <div
              class="v-image__image v-image__image--preload v-image__image--cover"
              style="background-position: center center;"
            />
            <div
              class="v-responsive__content"
            />
          </div>
        </a>
      </span>
    </div>
     
    <div
      class="col-sm-3 col"
    >
      <a
        class="recent-prod-link"
      >
        <div
          class="v-responsive view-more"
        >
          <div
            class="v-responsive__sizer"
            style="padding-bottom: 56.25%;"
          />
          <div
            class="v-responsive__content"
          >
            <h3>
              View More
            </h3>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;
